lib/backend/dynamo/dynamodbbk.go

```diff
@@ -617,44 +662,57 @@ const (
 
 // DeleteRange deletes range of items with keys between startKey and endKey
 func (b *Backend) DeleteRange(ctx context.Context, startKey, endKey backend.Key) error {
-	if startKey.IsZero() {
-		return trace.BadParameter("missing parameter startKey")
-	}
-	if endKey.IsZero() {
-		return trace.BadParameter("missing parameter endKey")
-	}
-	// keep fetching and deleting until no records left,
-	// keep the very large limit, just in case if someone else
-	// keeps adding records
-	for i := 0; i < backend.DefaultRangeLimit/100; i++ {
-		result, err := b.getRecords(ctx, prependPrefix(startKey), prependPrefix(endKey), batchOperationItemsLimit, nil)
+	// Attempt to pull all existing items and delete them in batches
+	// in accordance with the BatchWriteItem limits. There is a hard
+	// cap on the total number of items that can be deleted in a single
+	// DeleteRange call to avoid racing with additional records being added.
+	const maxDeletions = backend.DefaultRangeLimit / 100
+	requests := make([]types.WriteRequest, batchOperationItemsLimit)
+	var pageCount, totalCount int
+	for item, err := range b.Items(ctx, backend.IterateParams{StartKey: startKey, EndKey: endKey}) {
```

Comment:
We probably don't `DeleteRange` enough to matter, but we could avoid having to deal with the whole item if we queried with a projection expression that only had the fullpath in it.

We should consider doing that as a byproduct of expanding the backend interface to include a `Keys()` iterator, if we ever get there (having a `Count()` primitive would also be cool).